test/Microsoft.Azure.Devices.Edge.Test.Common/LeafDevice.cs

```diff
@@ -0,0 +1,352 @@
+                        token);
+                });
+        }
+
+        static async Task<Device> GetEdgeDeviceIdentityAsync(string parentId, IotHub iotHub, CancellationToken token)
+        {
+            Device edge = await iotHub.GetDeviceIdentityAsync(parentId, token);
+            if (edge == null)
+            {
+                throw new ArgumentException($"Device '{parentId}' not found in '{iotHub.Hostname}'");
+            }
+
+            return edge;
+        }
+
+        static async Task<LeafDevice> DeleteIdentityIfFailedAsync(Device device, IotHub iotHub, CancellationToken token, Func<Task<LeafDevice>> what)
+        {
+            try
+            {
+                return await what();
+            }
+            catch
+            {
+                await DeleteIdentityAsync(device, iotHub, token);
+                throw;
+            }
+        }
+
+        static async Task<LeafDevice> CreateLeafDeviceAsync(Device device, Func<DeviceClient> clientFactory, IotHub iotHub, CancellationToken token)
+        {
+            DeviceClient client = clientFactory();
+            await client.SetMethodHandlerAsync("DirectMethod", DirectMethod, null, token);
+            return new LeafDevice(device, client, iotHub);
+        }
+
+        public Task SendEventAsync(CancellationToken token)
+        {
+            var message = new Message(Encoding.ASCII.GetBytes(this.device.Id))
+            {
+                Properties = { ["leaf-message-id"] = this.messageId }
+            };
+            return this.client.SendEventAsync(message, token);
+        }
+
+        public Task WaitForEventsReceivedAsync(DateTime seekTime, CancellationToken token)
+        {
+            return Profiler.Run(
+                () => this.iotHub.ReceiveEventsAsync(
+                    this.device.Id,
+                    seekTime,
+                    data =>
+                    {
+                        data.SystemProperties.TryGetValue("iothub-connection-device-id", out object devId);
+                        data.Properties.TryGetValue("leaf-message-id", out object msgId);
+
+                        return devId != null && devId.ToString().Equals(this.device.Id)
+                                             && msgId != null && msgId.ToString().Equals(this.messageId);
+                    },
+                    token),
+                "Received events from device '{Device}' on Event Hub '{EventHub}'",
+                this.device.Id,
+                this.iotHub.EntityPath);
+        }
+
+        public Task InvokeDirectMethodAsync(CancellationToken token) =>
+            Profiler.Run(
+                () => this.iotHub.InvokeMethodAsync(
+                    this.device.Id,
+                    new CloudToDeviceMethod("DirectMethod"),
+                    token),
+                "Invoked method on leaf device from the cloud");
+
+        // BUG: callers can continue to try to use this object after DeleteIdentityAsync has been called.
```

Comment:
Should we implement IDisposable interface to call DeleteIdentityAsync or calling cleanup method?

Reply:
As mentioned earlier, managing the identity in test setup/teardown and making LeafDevice disposable are both problematic. I suppose if tests try to call LeafDevice methods after deleting the device's identity they'll get service errors, which is the expected behavior. This test code probably doesn't need to be resilient to bugs in the calling (test) code. I'll just delete this comment.

Reply:
Comment deleted.